pkg/certificate/configuration.go

```diff
@@ -2,169 +2,453 @@ package certificate
+	if err != nil && !new {
+		return err
+	}
+
+	objectOps := objectOperatorsMap[object.key.Kind]
+	objectOps.fromChainMapper(object, certificateChain)
+
+	if reflect.DeepEqual(old, object.kobject) {
+		// noop
+		return nil
+	}
+
+	if !reflect.DeepEqual(old, current) {
+		return fmt.Errorf("An object changed since originally read: %s", object.key)
+	}
+
+	if new {
+		logger.Info("Create object")
+		err = m.client.Create(context.TODO(), object.kobject)
+	} else {
+		logger.Info("Update object")
+		err = m.client.Update(context.TODO(), object.kobject)
+	}
+
+	return err
+}
+
+func initMutatingWebhook(name, namespace string) client.Object {
+	return &admissionregistrationv1.MutatingWebhookConfiguration{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+	}
+}
+
+func initValidatingWebhook(name, namespace string) client.Object {
+	return &admissionregistrationv1.ValidatingWebhookConfiguration{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+	}
+}
+
+func initSecret(name, namespace string) client.Object {
+	return &corev1.Secret{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Annotations: map[string]string{
+				secretManagedAnnotationKey: "",
+			},
+		},
+	}
+}
+
+// mapWebhookToChain maps a webhook object to certificate chain data.
+// If the webhook has no relevant data or does not exist, it is removed
+// from the object map so that is no longer considered. For every backing
+// service of the webhook, a reference to the service secret is added to the
+// object map if not already there.
+func mapWebhookToChain(object *keyedObject, objects objectMap, certificateChain *chain.CertificateChainData) {
+	clientConfigMap := clientConfigMap(object.kobject)
+	if len(clientConfigMap) <= 0 {
+		delete(objects, object.key)
+		return
+	}
+
+	if certificateChain.CertificatesIssued == nil {
+		certificateChain.CertificatesIssued = map[string]*chain.CertificateIssue{}
+	}
+
+	for name, config := range clientConfigMap {
+		serviceName := config.Service.Name
+		serviceNamespace := config.Service.Namespace
+		serviceHostname := serviceHostname(serviceName, serviceNamespace)
+
+		if _, found := certificateChain.CertificatesIssued[serviceHostname]; !found {
+			certificateChain.CertificatesIssued[serviceHostname] = newCertificateIssue(serviceName, serviceNamespace)
 		}
 
-		for _, clientConfig := range m.clientConfigList(webhook) {
-			// Update the CA bundle at webhook
-			updatedCABundle, err := updateCABundle(clientConfig.CABundle)
-			if err != nil {
-				return errors.Wrap(err, "failed updating CA bundle")
-			}
-			clientConfig.CABundle = updatedCABundle
+		caBundleName := caBundleName(object.key.String(), name)
+		certificateChain.CertificatesIssued[serviceHostname].CACertPEM[caBundleName] = config.CABundle
+		key := newObjectKey("Secret", serviceNamespace, serviceName)
+		if _, found := objects[key]; !found {
+			objects[key] = &keyedObject{key, nil}
```

Comment:
Why don't we call here the "readObjectToChain" so we don't need the for loop at "readObjectsToChain", I know it will be hidding recursiveness but the level or recursion here is super low.